src/common/plugins.js

```diff
@@ -0,0 +1,118 @@
+'use strict'
+
+const { app, nativeImage } = require('electron')
+const { join } = require('path')
+const { warn, verbose, logger } = require('./log')
+const { uniq, pluck } = require('./util')
+const { promises: jsonld } = require('jsonld')
+
+class Plugins {
+  constructor(root, plugins = []) {
+    this.root = root
+    this.plugins = plugins
+    this.instances = []
+    this.loadPaths = [
+      join(this.root, 'node_modules'),
+      this.root
+    ]
+    this.isRenderer = process.type === 'renderer'
+  }
+
+  get context() {
+    return {
+      logger,
+      jsonld,
+      nativeImage
+    }
+  }
+
+  contract(Plugin, config) {
+    return new Plugin(config, this.context)
+  }
+
+  get packages() {
+    return uniq(this.plugins.map(p => p.package))
+  }
+
+  // try to load the package from one of the paths
+  loadPackage(name) {
+    let pkg, hooks
+    for (let i = 0; i < this.loadPaths.length; i++) {
+      const path = join(this.loadPaths[i], name)
+      try {
+        pkg = require(path)
+        hooks = require(join(path, 'package.json')).hooks || {}
+      } catch (error) {
+        continue
+      }
+      if (pkg) return { pkg, hooks }
+    }
+    warn(`Plugin package "${name}" can not be loaded`)
+  }
+
+  initialize() {
+    // for each instance in `this.plugins`, require their package
+    // and store it in `this.instances`
+    for (let i = 0; i < this.plugins.length; i++) {
+      const params = this.plugins[i]
+
+      const result = this.loadPackage(params.package)
+      if (!result) continue
+
+      const { pkg, hooks } = result
+      this.instances.push({
+        params,
+        instance: this.isRenderer ? this.contract(pkg, params.config) : null,
```

Comment:
Instead of using extra `isRenderer` state here, let's add a parameter to `initialize` that decides whether or not the plugin should be instantiated (it's OK for the parameter to default to the true/false based on the process type).